src/GitExtensions.Extensibility.Tools/tools/Download-GitExtensions.ps1

```diff
@@ -0,0 +1,179 @@
+param(
+  [Parameter(Mandatory=$true)]
+  [string] $ExtractRootPath, 
+  [Parameter(Mandatory=$true)]
+  [string] $Version, 
+  [ValidateSet('GitHub','AppVeyor', ignorecase=$False)]
+  [string] $Source = "GitHub"
+)
+
+function Test-LocalCopy 
+{
+	Param(
+		[Parameter(Mandatory=$true, Position=0)]
+		[string] $ExtractPath,
+		[Parameter(Mandatory=$true, Position=1)]
+		[string] $FileName
+	)
+
+	$FilePath = [System.IO.Path]::Combine($ExtractPath, $FileName);
+	if (Test-Path $FilePath)
+	{
+		Write-Host "Download '$FileName' already exists.";
+		return $true;
+	}
+	
+	return $false;
+}
+
+function Find-ArchiveUrl 
+{
+	param (
+		[Parameter(Mandatory=$true, Position=0)]
+		[string] $Version,
+		[Parameter(Mandatory=$true, Position=1)]
+		[ValidateSet('GitHub','AppVeyor', ignorecase=$False)]
+		[string] $Source
+	)
+	
+	Write-Host "Searching for GitExtensions release on '$Source'.";
+	if ($Source -eq "github")
+	{
+		return Find-ArchiveUrlFromGitHub -Version $Version;
+	}
+	
+	if ($Source -eq "appveyor")
+	{
+		return Find-ArchiveUrlFromAppVeyor -Version $Version;
+	}
+
+	throw "Unable to find download URL for 'Git Extensions $Version'";
+}
+
+function Find-ArchiveUrlFromGitHub 
+{
+	param (
+		[Parameter(Mandatory=$true, Position=0)]
+		[string] $Version
+	)
+	
+	$Releases = Invoke-RestMethod -Uri 'https://api.github.com/repos/gitextensions/gitextensions/releases';
+	foreach ($Release in $Releases)
+	{
+		if ($Release.tag_name -eq $Version)
+		{
+			Write-Host "Selected release '$($Release.name)'.";
+			foreach ($Asset in $Release.assets)
+			{
+				if ($Asset.content_type -eq "application/zip" -and $Asset.name.Contains('Portable'))
+				{
+					Write-Host "Selected asset '$($Asset.name)'.";
+					return $Asset.browser_download_url;
+				}
+			}
+		}
+	}

```

Comment:
When I mentioned PS purist, I meant the script could be written in a PS- (or bash-) way. Again it doesn't mean it will be better, it's just different.

The whole things can be simplified to two lines:
```
$Releases = Invoke-RestMethod -Uri 'https://api.github.com/repos/gitextensions/gitextensions/releases';
$Releases.assets.browser_download_url | `
    Where-Object { $_.Contains("download/$Version/") -and $_.Contains('Portable') }
```
(took me a lazy hour while watching a TV)

Reply:
Yes, I got that. I was asking if one can print those "Selected asset..." messages in this syntax.

Reply:
```
$asset = $Releases.assets.browser_download_url | `
            Where-Object { $_.Contains("download/$Version/") -and $_.Contains('Portable') };
Write-Host "Selected asset: $asset";
```